### PR TITLE
HPCC-18389 Implement begin and end on hash tables

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [x] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith @richardkchapman please review.
This is the for syntax for the hash tables which gives the most benefit, and was fully implemented - I'll return to arrays later.
This function will not work very well if the hash table is being modified (neither would the previous one, but it will behave differently).  It should be more efficient though - since it directly accesses the elements by index.

File: system/jlib/jsuperhash.cpp [hunk 1]
```diff
@@ -448,6 +448,13 @@ bool SuperHashTable::removeExact(void *et)
     return true;
 }
 
+unsigned SuperHashTable::validIdx(unsigned i) const
+{
+    while (i < tablesize && !table[i])
+        i++;
+    return i;
```
Comment (jakesmith) [inline system/jlib/jsuperhash.cpp]: @ghalliday - What happens if there is no valid i ? e.g. no elements or beyond last? 
I didn't spot any error checking on begin() calling firstIdx() ..

Comment (ghalliday) [reply system/jlib/jsuperhash.cpp]: the maximum i can be is tablesize.  That should be the value of hash.end() - so it will not iterate through anything.

**Comment by richardkchapman** [comment]
@jakesmith

**Comment by jakesmith** [comment]
I was little unsure about e.g. ConstHashItem where HT had 0 elements, but I see begin() and end() can create ConstHashItem items even if 0 elements now and 'item' will not be called.

@ghalliday - looks good to merge.

**Pull request merged**